Prepare to stop using --whole-archive for libc-associated libraries

keith-packard opened the pull request:
Here are a set of small build patches which will make merging #65434 easier. Only one of them should cause any change in the resulting binaries (allowing the libc partition to be empty). #65434 has been rebased on top of this list.

keith-packard committed:
libc/picolibc: Split hooks into separate files

This splits the picolibc helper functions into separate files so that
code not explicitly referenced can be excluded from the result once
we stop using --whole-archive with this library.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/picolibc: Add 'write' implementation when needed

This is needed for the picolibc __chk_fail implementation as that
uses it up through version 1.8.5. After that, picolibc uses puts.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
kernel: Add kernel library early in the build

Create the kernel library object before loading in the subdirectories so
that other parts of the code (the toolchain and C library) may adjust the
kernel dependencies as needed using target_link_library(kernel, ...)

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
kernel/mem_domain: Allow z_libc_partition to be empty

We can't know before building the system whether there will be any data
present in the C library parition. Allow this, simply skipping any use of
it.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/common: Add toolchain.h to gmtime_r for ZRESTRICT

When built as a regular library, this file doesn't get included
automatically.

Signed-off-by: Keith Packard <keithp@keithp.com>

dcpleung commented on kernel/mem_domain.c:
```diff
@@ -168,6 +168,11 @@ int k_mem_domain_add_partition(struct k_mem_domain *domain,
 	}
 
+#ifdef Z_LIBC_PARTITION_EXISTS
+	if (part->size == 0 && part == &z_libc_partition)
+		goto out;
```
I think this should return an error instead.

keith-packard replied:
The reason for this change is to make having an empty libc partition *not* be an error. Without this change, it would be an error to have a zero sized libc partition already.

Predicting when the libc partition will be empty is going to be tricky when malloc is no longer always included. The malloc static variables live in the libc partition but they only land in the binary if there is a call to malloc somewhere in the application. And that can depend on a wide range of subsystem versions and configuration options.

keith-packard committed:
newlib: Provide _unlink when CONFIG_FILE_SYSTEM is set

_unlink is used by newlib's 'remove' implementation, which is
part of the standard C API, not a POSIX extension.

Signed-off-by: Keith Packard <keithp@keithp.com>

aescolar commented on lib/libc/picolibc/CMakeLists.txt:
```diff
@@ -1,7 +1,15 @@
+  )
 
 # Do not allow LTO when compiling libc-hooks.c file
 set_source_files_properties(libc-hooks.c PROPERTIES COMPILE_OPTIONS $<TARGET_PROPERTY:compiler,prohibit_lto>)
```
I guess this needs changing now?

keith-packard replied:
Thanks, yes it does.

aescolar commented on lib/libc/picolibc/CMakeLists.txt:
```diff
@@ -1,7 +1,15 @@
 zephyr_library()
-zephyr_library_sources(libc-hooks.c)
+zephyr_library_sources(
+  assert.c
```
A nit: Before from the file name it was clear what this was. For the shake of ease of understanding of whomever wanders thru here, would it make sense to either put this in a hooks subfolder or add a comment here and/or in the files?

keith-packard replied:
Hrm. These are *all* hooks required to map between Zephyr and picolibc; that's what this directory does.

aescolar reviewed (commented):
Note the lto line.
For the cmake change it would be very good if @tejlmand gave his blessing.

keith-packard commented:
> Note the lto line. For the cmake change it would be very good if @tejlmand gave his blessing.

I just removed the code disabling LTO -- it was inherited from the newlib porting layer; I'm not sure what it was trying to accomplish, but LTO should be fine for all of this code.

keith-packard committed:
newlib: provide _unlink_r for xtensa

This is required for newlib's remove implementation.

Signed-off-by: Keith Packard <keithp@keithp.com>

cfriedt commented on lib/libc/picolibc/write.c:
```diff
@@ -0,0 +1,21 @@
+#include <zephyr/sys/libc-hooks.h>
+
+/* Don't replace the version supplied by fdtable */
+#if !defined(CONFIG_FDTABLE) || !defined(CONFIG_POSIX_API)
```
At some point, because `CONFIG_POSIX_API` will be deprecated.

Would it make sense to have Picolibc select `CONFIG_POSIX_DEVICE_IO` and  `CONFIG_TC_PROVIDES_POSIX_DEVICE_IO` since it requires `write()` and maybe other symbols? They would still be provided by picolibc.

There would be no need for the ifdefs at that point.

Just a comment though, no changes needed currently.

tejlmand commented on CMakeLists.txt:
```diff
@@ -580,6 +580,21 @@ if(EXISTS ${APPLICATION_SOURCE_DIR}/VERSION)
+# directory, kernel/libkernel.a, link it with the application code instead
+# of building from source.
+
+set(libkernel ${CMAKE_CURRENT_SOURCE_DIR}/kernel/libkernel${CMAKE_STATIC_LIBRARY_SUFFIX})
+if(EXISTS ${libkernel})
+  add_library(kernel STATIC IMPORTED GLOBAL)
+  set_property(TARGET kernel PROPERTY IMPORTED_LOCATION ${libkernel})
+else()
+  add_library(kernel)
+endif()
```
i'm currently cleaning up Zephyr link handling, so I would like to know more details on the background reasons for this move.

The commit message gives a small hint, but not enough for me to actually be able to judge if the upcomming build system improvements are actually solving the issues faced.

Can you provide some more details / examples on what exactly this move will solve ?

keith-packard replied:
This creates the kernel library early in the build process so that dependency information can be associated with it by early libraries in the system. In the follow-on patch, which actually makes the C library support bits into regular libraries, we need to construct a circular dependency graph including the kernel and C libraries. Doing that in the C library code instead of the kernel code seemed a bunch cleaner to me, but that required having the kernel library created before the C library support bits were processed by cmake.

tejlmand reviewed (commented):
Nice cleanup, only need some more details.
